### DISCUSSION TITLE
[Enhancement]: allow using pr_id as a Preview Deployments Environment Variables

### DISCUSSION THREAD

**Discussion by larbijirari** [coollabsio/coolify#4757] [Improvement Requests]
### Request Type

Improvement

### Description

this will help linking the backend and frontend apps.
example: 
where $pr_id is representing the current **pr_id**
![Image](https://github.com/user-attachments/assets/68609457-5a88-49d7-a575-04c0a07235f4)

**Comment by MatanMaimon** [comment]
Any update on this one!? 🙏

**Comment by nil0ft** [comment]
Would like to see that!

**Comment by Arnaud-IFEA** [comment]
Hello,
i don't even understand how people can make the preview deployment to work actually !
if you have an url in your envvar it can't adapt to a preview deployment.
If with docker compose you use a service name to connect to the database (you don't want a public URL!), how can you make it work knowing that preview deployment changes service names.

**Comment by astabeno** [comment]
I worked for a while trying to get preview branches to work.  It seems like this feature is not really ready unless you have a pretty simple setup.  A docker-compose with multiple containers and a tightly coupled DB that needs isolation in PR environment doesn't seem feasible yet.

**Comment by atlijohanns** [comment]
This would be a great! Makes syncing two preview apps in a monorepo so much simpler.

**Comment by faahim** [comment]
Upvoting this! ⬆️

**Comment by bysontin** [comment]
Upvoting this too. It would be a great enhancement !

**Comment by djsisson** [comment]
if someone wants to test, try add the following (it has to start with COOLIFY):

```php
            if ($this->application->environment_variables_preview->where('key', 'COOLIFY_PR_ID')->isEmpty()) {
                  $coolify_envs->put('COOLIFY_PR_ID', $this->pull_request_id);
            }
```

to here

https://github.com/coollabsio/coolify/blob/723f14e6f6a84bab744a792c3e628c5e5a3fa308/app/Jobs/ApplicationDeploymentJob.php#L1621-L1647

**Reply by atlijohanns** [reply]
Hey @djsisson. 

Could you help me understand where this code is located? Is it somewhere in the Coolify install on my underlying self hosted environment?

**Reply by djsisson** [reply]
although you can exec into your docker container and manually edit the php files and then update using some artisan commands, i don't recommend it

i meant the above for testing locally, the file is linked above the bottom code section

**Comment by nemvince** [comment]
+1

**Comment by chrisgen19** [comment]
any status on this?

**Comment by ondrejznojil** [comment]
+1

Would really love to see this feature. Otherwise, it is very hard to use preview deployments.

**Comment by Cinzya** [comment]
What's the usecase / reason that the PR ID is needed as a env variable?

If it's to pass the URL of a preview deployment, you can use Coolifys Predefined environment variables. For Applications that will be `COOLIFY_URL` and for service stacks `SERVICE_URL_IDENTIFIER`.
There is also a `SERVICE_NAME_IDENTIFIER` which resolves to the name of a service in a service stack.

**Reply by matheusfillipe** [reply]
> What's the usecase / reason that the PR ID is needed as a env variable?
> 
> If it's to pass the URL of a preview deployment, you can use Coolifys Predefined environment variables. For Applications that will be `COOLIFY_URL` and for service stacks `SERVICE_URL_IDENTIFIER`. There is also a `SERVICE_NAME_IDENTIFIER` which resolves to the name of a service in a service stack.

For docker compose, if I have an API service and a frontend static page that is built to use that API. Both urls will be dynamic and use the pr_id number. Is there a way I can reference the right API preview service url to have the preview for the whole stack?

**Reply by Cinzya** [reply]
yes. Lets say your compose looks like this
```
frontend:
	image: nginx:latest
...
api:
	image: node:latest
...
```
then you can get the url for the frontend service via the predefined environment variables `SERVICE_URL_FRONTEND` and for the api `SERVICE_URL_API`.

**Reply by SuperRonanCraft** [reply]
I see a problem about this as some libraries especially @t3-oss/env-nextjs a typesafe environment variable library expect environment variables, that are meant for client's, require to be prefixed with "NEXT_PUBLIC" else it literally doesn't let you create a client environment variable. 
I actually need the `{{pr_id}}` to be available to I can create a preview env variable called `NEXT_PUBLIC_FRONTEND_URL=https://{{pr_id}}.preview.myapp.com` so I can give clients the current frontend url that I described, also another example would be the auth call back url like `NEXTAUTH_URL` from NextAuth, it would allow me to have dynamic auth routes so authentication isn't broken between pr's
Since the pr creates a new url each time, I cant use just "preview.myapp.com" because my team works on multiple pr's at one time so I need the flexibility, if this pr variable can available for environment variables it would make github pre-deployments a dream!

**Reply by Cinzya** [reply]
What stops you from doing
```
NEXT_PUBLIC_SERVICE_URL_FRONTEND=$SERVICE_URL_FRONTEND
```
?

**Reply by kareemmahlees** [reply]
@Cinzya Here is a use case I am currently encountering and don't know how to get about it.

I am currently trying to setup preview deployments with each deployment connecting to a separate database branch, I would have  liked to be able to add the following env var to my preview deployments:
```
DATABASE_URL=postgresql://postgres:postgres@localhost:5432/$pr_id
```
and in my CI/CD I would create a database branch with that PR id or number before triggering a deployment in coolify.

Can this be done now with the available env vars or this is a valid use-case where some PR specific env vars would have solved it?

**Reply by LeanderM99** [reply]
@kareemmahlees Im currently having the same issue. Could you please inform me if you find a solution for that problem?

**Reply by brandonjp** [reply]
@Cinzya is there an existing way to accomplish what @kareemmahlees is trying to do? I'm also trying to get something like:
`DATABASE_URL=postgres://app_preview:abc@xyz:555/db_pr_${PR_ID}` working

**Reply by Cinzya** [reply]
@brandonjp depending on what Build Pack you use, you could either use the `SERVICE_NAME_IDENTIFIER` or `COOLIFY_CONTAINER_NAME` env variable. In case of a docker compose, e.g.:

```yaml
frontend:
	image: nginx:latest
...
api:
	image: node:latest
...
```
`SERIVCE_NAME_FRONTEND` and `SERVICE_NAME_API` will return the value `frontend-pr-1` and `api-pr-1`. Or `COOLIFY_CONTAINER_NAME` which will work on Dockerfile / Nixpacks builds, returning `<coolify-container-name>-pr-1`.
Given you don't mind the service / container name being part of the database name.

**Reply by lukasgjetting** [reply]
@Cinzya Is there a way to reference other service urls/names without Docker Compose (using raw Dockerfiles)?

Specifically, I am trying to add an environment variable `API_URL` to my `web` service. `api` previews are deployed to `{{pr_id}}.api.dev.example.com` and from what I can tell, there is no easy way to bind that in the environment variables for `web`?

 Ideally it would look something like:

<img width="665" height="448" alt="image" src="https://github.com/user-attachments/assets/bfc12014-97aa-42e6-8197-fd947d6d2e74" />

Or `$SERVICE_URL_WEB` perhaps, so it matches the Docker Compose convention? 

Assuming all services follow the same domain pattern, it's technically possible for each to parse their own url to find the PR number. It's not very nice/ergonomic, though.

Thanks!

**Reply by Cinzya** [reply]
@lukasgjetting the URL of your own app, when it's not a docker compose, you can get via the `COOLIFY_URL` variable. 
If you need to URL of a different reaource, you could call the Coolify API and gather that during runtime.

We have a list the predefined variables in the docs: https://coolify.io/docs/applications/configuration/environment-variables#predefined-application-variables

**Reply by lukasgjetting** [reply]
@Cinzya Yeah that's what I concluded as well.

In that case, I think a predefined pr number env var would make a lot of sense, since it would allow all the config to happen at build time instead of having each app look up the sibling service URLs at runtime.

I am admittedly not too familiar with the codebase, but it seems like it would be a pretty easy add in [app/Jobs/ApplicationDeploymentJob.php](https://github.com/coollabsio/coolify/blob/7b84e8722c52fa2a9bfb5dc331b9eeb908402723/app/Jobs/ApplicationDeploymentJob.php#L3211) to do something like:

```php
if ($this->application->environment_variables_preview
    ->where('key', 'COOLIFY_PR_NUMBER')->isEmpty()) {
    $coolify_envs->put(
        'COOLIFY_PR_NUMBER',
        (string) $this->pull_request_id
    );
}
```

Caching shouldn't be affected either, since we already inject `COOLIFY_BRANCH` in the same path.

Would the team be open to a PR if I were to implement and test it? Would be happy to put in the work. :)

Thanks!

**Comment by hubertoschusch** [comment]
Bump
